src/instantx.cpp

```diff
@@ -56,304 +52,323 @@ void ProcessMessageInstantSend(CNode* pfrom, std::string& strCommand, CDataStrea
+                    // from the same masternode.
+                    // TODO: apply pose ban score to this masternode?
+                    // NOTE: if we decide to apply pose ban score here, this vote must be relayed further
+                    // to let all other nodes know about this node's misbehaviour and let them apply
+                    // pose ban score too.
+                    LogPrintf("CInstantSend::ProcessTxLockVote -- masternode sent conflicting votes! %s\n", vote.GetMasternodeOutpoint().ToStringShort());
+                    return false;
+                }
+            }
         }
-        return false;
+        // we have votes by other masternodes only (so far), let's continue and see who will win
+    } else {
+        std::set<uint256> setHashes;
+        setHashes.insert(txHash);
+        mapVotedOutpoints.insert(std::make_pair(vote.GetOutpoint(), setHashes));
     }
 
-    if(!mapTxLockCandidates.count(vote.txHash)) {
+    if(!it->second.AddVote(vote)) {
         // this should never happen
         return false;
     }
 
-    //compile consessus vote
-    mapTxLockCandidates[vote.txHash].AddVote(vote);
-
-    int nSignatures = mapTxLockCandidates[vote.txHash].CountVotes();
-    LogPrint("instantsend", "ProcessTxLockVote -- Transaction Lock signatures count: %d, vote hash=%s\n", nSignatures, vote.GetHash().ToString());
-
-    if(nSignatures >= INSTANTSEND_SIGNATURES_REQUIRED) {
-        LogPrint("instantsend", "ProcessTxLockVote -- Transaction Lock Is Complete! txid=%s\n", vote.txHash.ToString());
-
-        if(!FindConflictingLocks(mapLockRequestAccepted[vote.txHash])) { //?????
-            if(mapLockRequestAccepted.count(vote.txHash)) {
-                UpdateLockedTransaction(mapLockRequestAccepted[vote.txHash]);
-                LockTransactionInputs(mapLockRequestAccepted[vote.txHash]);
-            } else if(mapLockRequestRejected.count(vote.txHash)) {
-                ResolveConflicts(mapLockRequestRejected[vote.txHash]); ///?????
-            } else {
-                LogPrint("instantsend", "ProcessTxLockVote -- Transaction Lock is missing! nSignatures=%d, vote hash=%s\n", nSignatures, vote.GetHash().ToString());
+    int nSignatures = it->second.CountVotes();
+    int nSignaturesMax = it->second.txLockRequest.GetMaxSignatures();
+    LogPrint("instantsend", "CInstantSend::ProcessTxLockVote -- Transaction Lock signatures count: %d/%d, vote hash=%s\n",
+            nSignatures, nSignaturesMax, vote.GetHash().ToString());
+
+    if(it->second.IsAllOutPointsReady()) {
+        // we have enough votes now
+        LogPrint("instantsend", "CInstantSend::ProcessTxLockVote -- Transaction Lock Is Complete! txid=%s\n", txHash.ToString());
+        if(ResolveConflicts(it->second)) {
+            // the order here matters
+            UpdateLockedTransaction(it->second);
+            if(!IsLockedInstantSendTransaction(txHash)) {
```

Comment:
If the tx is already locked, it seems like we should return earlier than this (ie. why call ResolveConflicts/UpdateLockedTransaction ?).